wallet-new/src/Cardano/Wallet/Kernel/DB/InDb.hs

```diff
@@ -7,99 +8,771 @@ module Cardano.Wallet.Kernel.DB.InDb (
+    InDb (a :: Core.Attributes Core.AddrAttributes) <- SC.safeGet
+    InDb (t :: Core.AddrType) <- SC.safeGet
+    pure (InDb (Core.Address r a t))
+  putCopy (InDb (Core.Address r a t)) = SC.contain $ do
+    SC.safePut (InDb (r :: Core.AddressHash Core.Address'))
+    SC.safePut (InDb (a :: Core.Attributes Core.AddrAttributes))
+    SC.safePut (InDb (t :: Core.AddrType))
+
+instance SC.SafeCopy (InDb Core.AddrAttributes) where
+  getCopy = SC.contain $ do
+    yiap :: Maybe (InDb Core.HDAddressPayload) <- SC.safeGet
+    InDb (ast :: Core.AddrStakeDistribution) <- SC.safeGet
+    pure (InDb (Core.AddrAttributes (fmap _fromDb yiap) ast))
+  putCopy (InDb (Core.AddrAttributes yap asr)) = SC.contain $ do
+    SC.safePut (fmap InDb yap)
+    SC.safePut asr
+
+instance SC.SafeCopy (InDb Core.AddrStakeDistribution) where
+  getCopy = SC.contain $ fmap InDb $ do
+    SC.safeGet >>= \case
+      0 -> pure Core.BootstrapEraDistr
+      1 -> Core.SingleKeyDistr <$> fmap _fromDb SC.safeGet
+      2 -> Core.UnsafeMultiKeyDistr <$> fmap _fromDb SC.safeGet
+      (_ :: Word8) -> empty
+  putCopy (InDb x) = SC.contain $ case x of
+    Core.BootstrapEraDistr -> SC.safePut (0 :: Word8)
+    Core.SingleKeyDistr a -> do
+      SC.safePut (1 :: Word8)
+      SC.safePut (InDb (a :: Core.StakeholderId))
+    Core.UnsafeMultiKeyDistr m -> do
+      SC.safePut (2 :: Word8)
+      SC.safePut (InDb (m :: Map Core.StakeholderId Core.CoinPortion))
+
+instance SC.SafeCopy (InDb Core.CoinPortion) where
+  getCopy = SC.contain $ do
+    w :: Word64 <- SC.safeGet
+    pure (InDb (Core.CoinPortion w))
+  putCopy (InDb (Core.CoinPortion w)) = SC.contain $ do
+    SC.safePut (w :: Word64)
+
+instance SC.SafeCopy (InDb Core.HDAddressPayload) where
+  getCopy = SC.contain $ do
+    bs :: B.ByteString <- SC.safeGet
+    pure (InDb (Core.HDAddressPayload bs))
+  putCopy (InDb (Core.HDAddressPayload bs)) = SC.contain $ do
+    SC.safePut (bs :: B.ByteString)
+
+
+instance SC.SafeCopy (InDb Core.Coin) where
+    getCopy = SC.contain $ do
+       w :: Word64 <- SC.safeGet
+       pure (InDb (Core.Coin w))
+    putCopy (InDb (Core.Coin w)) = SC.contain $ do
+       SC.safePut (w :: Word64)
+
+instance SC.SafeCopy (InDb Core.SlotId) where
+    getCopy = SC.contain $ do
+      InDb (e :: Core.EpochIndex) <- SC.safeGet
+      InDb (s :: Core.LocalSlotIndex) <- SC.safeGet
+      pure (InDb (Core.SlotId e s))
+    putCopy (InDb (Core.SlotId e s)) = SC.contain $ do
+      SC.safePut (InDb e)
+      SC.safePut (InDb s)
+
+instance SC.SafeCopy (InDb Core.Timestamp) where
+  getCopy = SC.contain $ do
+    msi :: Integer <- SC.safeGet
+    let ms :: Data.Time.Units.Microsecond = fromInteger msi
+    pure (InDb (Core.Timestamp ms))
+  putCopy (InDb (Core.Timestamp ms)) = SC.contain $ do
+    let msi :: Integer = toInteger (ms :: Data.Time.Units.Microsecond)
+    SC.safePut msi
+
+instance SC.SafeCopy (InDb Txp.TxAux) where
+  getCopy = SC.contain $ do
+    InDb (tx :: Txp.Tx) <- SC.safeGet
```

Comment:
I also personally prefer 2 space indent, but this code ....

Reply:
Yeah, I'll get the style right before merging it all in. I'm mostly concerned about the correctness issues and failing tests atm